src/operator/communication/exchange.rs

```diff
@@ -174,8 +174,8 @@ where
 
         for receiver in 0..self.npeers {
             *self.mailbox(sender, receiver).lock().unwrap() = data.next();
-            self.sender_counters[sender].fetch_sub(1, Ordering::SeqCst);
-            let old_counter = self.receiver_counters[receiver].fetch_add(1, Ordering::SeqCst);
+            self.sender_counters[sender].fetch_sub(1, Ordering::Release);
```

Comment:
From the `.fetch_add()` docs:
> Note that using `Acquire` makes the store part of this operation `Relaxed`, and using `Release` makes the load part `Relaxed`.

This is fine, we don't use the loaded value

Reply:
We do use it.  We increment it and store it back.  If we load the wrong value, the stored value will be wrong too.

Reply:
Orderings influence which writes of other variables are visible to the thread doing this atomic op, but they don't make any write less atomic. The change is correct

Reply:
Yes, the operation is atomic.  Question is, in what order it appears relative to other operations.  In particular, it may observe and decrement an old value of the counter.  This might be ok, but it's not obvious to me, and this is the kind of bug that takes weeks to debug.

Reply:
Each atomic object has a single strict modification order, that's impossible

Reply:
> Subsequent loads of the counter will be ordered wrt this operation due to the use of `Release`, but the value they observe is not ordered wrt to earlier stores.

There will be a happens-before edge between the earlier writes to `mailbox(sender, receiver)` and the operations following an `Acquire` of `sender_counters` due to the **release sequence** headed by the last write to `sender_counters`.
(I also assumed for a long time that `AcqRel` was necessary for release sequences but it turns out it's not so; the only thing is whether the sequence is full of RMWs. So `Release` RMWs should be sufficient.)

Reply:
> Each atomic object has a single strict modification order, that's impossible

But this order may not be consistent with loads/stores to other locations.